File: js/lib/device.js
exports.Device = Device
function Device(family) {
  this.family = family || 'Other';
}

Device.prototype.toString = function() {
  return this.family;
};


exports.makeParser = function(regexes) {
  var parsers = regexes.map(function (obj) {
    var regexp = new RegExp(obj.regex),
        deviceRep = obj.device_replacement;

    function parser(str) {
      var m = str.match(regexp);
      if (!m) { return null; }

      var family = deviceRep ? deviceRep.replace('$1', m[1]) : m[1];
      return new Device(family);
    }

    return parser;
  });

  function parser(str, ua_family, os_family) {
    var obj;

    if (typeof str === 'string') {
      for (var i = 0, length = parsers.length; i < length; i++) {
        obj = parsers[i](str, ua_family, os_family);
        if (obj) { return obj; }
      }
    }

    return obj || new Device();
  }

  return parser;
};

File: go/uaparser/device.go
package uaparser

import (
	"regexp"
	"strings"
)

type Device struct {
	Family string
}

type DevicePattern struct {
	Regexp            *regexp.Regexp
	Regex             string
	DeviceReplacement string
}

func (dvcPattern *DevicePattern) Match(line string, dvc *Device) {
	bytes := dvcPattern.Regexp.FindStringSubmatch(line)
	if len(bytes) > 0 {
		groupCount := dvcPattern.Regexp.NumSubexp()

		if len(dvcPattern.DeviceReplacement) > 0 {
			if strings.Contains(dvcPattern.DeviceReplacement, "$1") && groupCount >= 1 && len(bytes) >= 2 {
				dvc.Family = strings.Replace(dvcPattern.DeviceReplacement, "$1", bytes[1], 1)
			} else {
				dvc.Family = dvcPattern.DeviceReplacement
			}
		} else if groupCount >= 1 {
			dvc.Family = bytes[1]
		}
	}
}

func (dvc *Device) ToString() string {
	return dvc.Family
}
